### DISCUSSION TITLE
Microcontrollers & Drones

### DISCUSSION THREAD

**Discussion by zack-emmert** [fnuecke/oc2#156] [Ideas]
Are there any plans to include something analogous to the microcontrollers & drones of OC1? I always found them to be very useful for small-scale projects that didn't require a full computer. And if there is, how would component interface work? I'm not sure how well suited a JSON RPC system is to embedded software.

**Comment by fnuecke** [comment]
Microcontrollers probably not, it'd more likely be a slimmer OS (I started dabbling with eLua, but burnt out on that), which would require a lot less ram.

Drones possibly, I've been thinking about how to integrate those; would probably be quite different to how they were before (e.g. not each their own VM/computer). No specific plans yet though.

**Reply by zack-emmert** [reply]
It might be interesting to have drones simply be a "dumb" entity controlled by a separate computer through, say, a drone controller card or something. Perhaps the drone could take upgrades similar to the robot, but let the computer control that functionality via the same controller card.

**Reply by SquidDev** [reply]
[Everything old is new again](http://www.computercraft.info/forums2/index.php?/topic/20528-quadracoptors-020-mc-1710-cc-165/) :p.

**Reply by SergSel2006** [reply]
If you integrate microcontroller you will eventually be able to integrate drone as it reuses much from microcontroller.
As for microcontroller... I would like to see limits increased. It's pointless to require end-game resources to run a computer (because lowest-tier RAM, for example, is actually pointless). My idea is:
Increase Flash memory storage to be at least 4MiB so you can actually run rootfs here, which is enough for a microcontroller.
Increase tier 1 RAM to 16MiB, tier 2 RAM to 32MiB and, eventually, tier 3 to 64MiB. 
Increase hard disks: Tier 1 to 8 MiB, Tier 2 to 16MiB and tier 3 to 32MiB. 
Increase Floppy to be at least 1 MiB.
This way everything seems more playable. 
For microcontroller: You can re-create assembler and microcontroller case from OC. The case will be able to hold tier 1 memory, tier 1 disk and 2 cards. With increased ROM we will be able to run most basic rootfs and 1 disk will give us space to store our data. The 32MB memory will allow us to run built-in linux actually.
The idea of increasing RAM and space is to make compile tools work here so we will be able to build our things inside the game, recreating coding side of original OpenComputers, and parallely making microcontroller working here. The downside of increased resources usage is something that is... Well, unavoidable if we need it for mod playability. If you don't accept my numbers for this, you can make it less, but we need to increase it anyways!
And the Flash memory must have some sort of flasher, so microcontrollers and drones can be actually programmed.
As for drones, we should see the microcontroller part working first.

**Reply by SergSel2006** [reply]
Another idea is to make Lua ROM. And then you can just re-create original stuff.

**Reply by JohnTheCoolingFan** [reply]
Maybe emulate something similar to a real microcontroller? AVR or STM32, for example.
